src/main/java/com/hms_networks/americas/sc/extensions/historicaldata/RapidCatchUp.java

```diff
@@ -0,0 +1,223 @@
+   * @param startTimeMilliseconds milliseconds since epoch UTC
+   * @return true/false should enter rapid catch up mode
+   */
+  public static boolean shouldEnterRapidCatchUpMode(
+      boolean lastReadEmpty, long startTimeMilliseconds) {
+
+    /* If last read to historical logs was empty and the current tracking time is older than RAPID_CATCH_UP_TIME_DELTA_MILLISECONDS, enable RapidCatchUp */
+    return lastReadEmpty
+        && System.currentTimeMillis() - startTimeMilliseconds
+            > RAPID_CATCH_UP_TIME_DELTA_MILLISECONDS;
+  }
+
+  /**
+   * Dynamically adjust the request period to by doubling it, but do not exceed {@link
+   * RAPID_CATCH_UP_TIME_DURATION_MAX}.
+   */
+  public static void adjustRapidCatchUpPeriod() {
+    catchUpRequestDurationMilliseconds =
+        Math.min(2 * catchUpRequestDurationMilliseconds, RAPID_CATCH_UP_TIME_DURATION_MAX);
+  }
+
+  /** Reset the request period to {@link RAPID_CATCH_UP_TIME_BASE}. */
+  public static void resetRapidCatchUpPeriod() {
+    catchUpRequestDurationMilliseconds = RAPID_CATCH_UP_TIME_BASE;
+  }
+
+  /**
+   * Special purpose stream reader to enable rapid catch up. Performs a limited number of reads and
+   * returns up to {@link MAX_STREAM_READS} from {@link InputStream} as a {@link String} in the
+   * specified encoding.
+   *
+   * @param inputStream input stream to read
+   * @param encoding encoding to use
+   * @return string contents of input stream (not complete)
+   * @throws IOException if an error occurs reading, closing the streams
+   * @throws NullPointerException should the input stream be a null
+   * @throws UnsupportedEncodingException when encoding is not supported
+   */
+  private static String getStringFromInputStreamCheckAvailable(
+      InputStream inputStream, String encoding)
+      throws IOException, NullPointerException, UnsupportedEncodingException {
+    // Create output stream for result
+    ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+
+    // Create buffer for reading input stream
+    byte[] buffer = new byte[STREAM_BUFFER_SIZE_BYTES];
+    int readCount = 0;
+    // Read from input stream and write to output stream
+    final int endOfStreamValue = -1;
+    final int offset = 0;
+    int length = inputStream.read(buffer);
+    while (length != endOfStreamValue && readCount < MAX_STREAM_READS) {
+      outputStream.write(buffer, offset, length);
+      readCount++;
+      length = inputStream.read(buffer);
+    }
+
+    // Get result from output stream and cleanup streams
+    String result = outputStream.toString(encoding);
+    outputStream.close();
+    inputStream.close();
+
+    return result;
+  }
+
+  /**
+   * Make rapid catch up request. This request does not reading and returning {@link DataPoint}. It
+   * checks to see if they existing in the historical time period, and if they do, return a tracker
+   * with the very first timestamp. If not, return a tracker with end time.
```

Comment:
Grammer needs to be fixed in this comment, here is an attempt.

```suggestion
   * Make rapid catch up request. This request does not return a {@link DataPoint}. It
   * checks to see if any {@link DataPoint} exist in the historical time period, and if they do, returns a tracker
   * with the very first timestamp. If not, return a tracker with end time.
```

Reply:
done